### PR TITLE
[Datsets] Add `SimpleTensorFlowDatasource`

### PR DESCRIPTION
**Pull request by bveeramani**
**IMPORTANT**: This PR is stacked on #23926. Merge that first.

## Why are these changes needed?

This PR makes it easier to use TensorFlow datasets with Ray Data.

## Related issue number

Towards #23977

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (bveeramani) [inline python/ray/data/datasource/tensorflow_datasource.py:13]: Similar to `SimpleTorchDatasource` (#23926), I'm not sure where to place this in the Ray Data documentation (if at all).

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (bveeramani) [inline python/ray/data/datasource/tensorflow_datasource.py:23]: Not if this line is needed, but this is what it looks like if I use my Python interpreter.

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (bveeramani) [inline python/ray/data/datasource/tensorflow_datasource.py:33]: Could use suggestions on how to simplify this part of the example.

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (bveeramani) [inline python/ray/data/datasource/tensorflow_datasource.py:47]: I can't raise a `ValueError` if the user passes a `tf.data.Dataset`, because Ray can't serialize a `tf.data.Dataset` in the first place.

File: python/ray/data/tests/test_dataset_formats.py
Comment (bveeramani) [inline python/ray/data/tests/test_dataset_formats.py:1481]: Could use suggestions on how to simplify this part of the test. I'm not super familiar with TensorFlow, so I figure there's a better API to use here.

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (matthewdeng) [reply python/ray/data/datasource/tensorflow_datasource.py:47]: Add a test case for the error that does get raised when trying to pass it in?

File: python/ray/data/tests/test_dataset_formats.py
Comment (matthewdeng) [reply python/ray/data/tests/test_dataset_formats.py:1481]: This looks fine to me.

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (matthewdeng) [inline python/ray/data/datasource/tensorflow_datasource.py:14]:
```suggestion
    """A datasource that lets you use TensorFlow datasets with Ray Data.
```

**Review by matthewdeng — COMMENTED** [review]
In general looks good to me!

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (clarkzinzow) [reply python/ray/data/datasource/tensorflow_datasource.py:33]: The example LGTM, what bits of it are you wanting to simplify?

**Comment by clarkzinzow** [comment]
Once the docs changes indicated [here](https://github.com/ray-project/ray/pull/23926#issuecomment-1105534061) are made and merged, we should similarly add this datasource to the API docs + add a usage example to the "Creating Datasets" guide.

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/tensorflow_datasource.py:47]: I don't think we shouldn't test that. 

We'd be testing that Ray Core can't serialize TensorFlow datasets. IMO, that's not a useful test.

File: python/ray/data/datasource/tensorflow_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/tensorflow_datasource.py:33]: Just thought the example felt a bit long. If it looks good to you, it's probably fine

File: python/ray/data/tests/test_dataset_formats.py
Comment (bveeramani) [reply python/ray/data/tests/test_dataset_formats.py]: Fixed

**Comment by clarkzinzow** [comment]
@bveeramani Same here, could you merge latest master? https://github.com/ray-project/ray/pull/23926#issuecomment-1111193345

**Review by clarkzinzow — APPROVED** [review]
LGTM after rebasing on/merging with the merged `TorchDatasource`!

**Comment by zhe-thoughts** [comment]
Marking as `release-blocker` for 1.13.0, where we plan to GA Datasets

**Comment by clarkzinzow** [comment]
Ah @bveeramani looks like `tensorflow_datasets` needs to be added as a test dependency, doing that now.

**Comment by clarkzinzow** [comment]
Datasets tests look good, merging!

**Pull request merged**
